Added parser that handles math expressions, strings, variables, and function-calls.

lukasberglund opened the pull request.

ayazhafiz commented:
Hi Lukas, I think we will need to bump up the test coverage on this commit (it is rather low, at around 88%). Can you take a look at the coverage report on coveralls and add more tests to improve the coverage? Lmk if you need any help

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+
+import Symbol
+
+
```
Extra whitespace? Can you run this through with [`hindent`](https://github.com/chrisdone/hindent)? I will add a CI test for this later on. Let me know if that sounds good or you would prefer something else.

lukasberglund replied:
Done.

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+
+
+
+data DataType = Int8 | Int16 | Int32 | Int64 | StandardInt | UInt8 | UInt16
```
What is a `StandardInt`? Why  not just `Int`?

lukasberglund replied:
You are right I'll just use that.

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+
+data DataType = Int8 | Int16 | Int32 | Int64 | StandardInt | UInt8 | UInt16
+                | UInt32 | UInt64 | UInt | Char
+                | Array {dataType :: DataType, size :: Int}
```
nice!

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+data DataType = Int8 | Int16 | Int32 | Int64 | StandardInt | UInt8 | UInt16
+                | UInt32 | UInt64 | UInt | Char
+                | Array {dataType :: DataType, size :: Int}
+                deriving (Show, Read, Eq)
```
to bump the test coverage, you can add tests that check that the derived versions of this act correctly

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+
+-- type Conditional = IfStatement | ElseStatement
+--
+-- data ElseStatement = ElseStatement
```
please remove

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+
+-- data Sign = Positive | Negative
+
+-- data Number = Integer Int
```
please remove

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+-- data ElseStatement = ElseStatement
+
+
+type BooleanExpression = Expression
```
so why have a separate type?

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+                  | FunctionCall String [Expression]
+                  | Conditional {condition :: Expression
+                                  , thenArm :: Expression
+                                  , elseArm :: Expression}
```
nice!

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+data Expression = InfixOperation Expression Operator Expression
+                  | PrefixOperation Operator Expression
+                  | Number Int | String String | Variable String
+                  | FunctionCall String [Expression]
```
Can you confirm that this means a function call is a string and an array of expressions?

lukasberglund replied:
It does.

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,36 @@
+data Expression = InfixOperation Expression Operator Expression
+                  | PrefixOperation Operator Expression
+                  | Number Int | String String | Variable String
+                  | FunctionCall String [Expression]
```
Would it be possible to define each of these data as separate data types? So that we have the opportunity to plug in + plug out as needed.

lukasberglund replied:
What do you mean by plug in + plug out?

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,143 @@
+module Parser where
+
+import qualified Symbol (Symbol(..))
```
consider

```haskell
import qualified Symbol
import Symbol (Symbol(..))
```

This will allow you to use `Symbol` types with the `Symbol` namespace without having to namespace the ctor as well.

lukasberglund replied:
I'm confused as to what this means. Are you saying that this would allow me to use things from Symbol without having to write Symbol.Whatever? If so, couldn't I just do
```haskell
import Symbol
```
to do that?

Also, I am doing import qualified because some names from Symbol clash with names from Grammar.

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,143 @@
+
+import qualified Symbol (Symbol(..))
+import Token (Token(..))
+import qualified Grammar
```
consider import `Expression` unqualified

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,143 @@
+parsePrefix :: [Token] -> (Grammar.Expression, [Token])
+-- parsePrefix - Parse operations that do not depend on previous symbols
+--               these include prefix operations like +1 or -7, but also
+--               function calls like
```
like?

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,143 @@
+
+operatorPrecidence :: Grammar.Operator -> Int
+operatorPrecidence op = case op of
+  Grammar.Paren -> -1
```
why negative?

lukasberglund replied:
Because the default call to parseExpression is with a precedence of zero, but I want parseExpression to stop when it sees a closing parenthesis, hence I made it zero. If you prefer I can  shift those numbers up :)

ayazhafiz replied:
no, thats a great reason

ayazhafiz commented on test/Parser/ParserSpec.hs:
```diff
@@ -0,0 +1,105 @@
+-- import qualified Builtin
+-- import Lexer (lex)
+-- import Symbol (Symbol(..))
+-- import Token (Token(..))
```
please remove

ayazhafiz commented on test/Parser/ParserSpec.hs:
```diff
@@ -0,0 +1,105 @@
+spec = do
+  describe "parseExpression" $ do
+    it "parses simple prefix expressions correctly" $ do
+      parseExpression (lex "+2") `shouldBe`
```
consider defining a helper function that will call `lex` and `parseExpression` for you.

lukasberglund replied:
Where do you think I should define it?

ayazhafiz replied:
At the top of this test file.

ayazhafiz commented on test/Parser/ParserSpec.hs:
```diff
@@ -0,0 +1,105 @@
+    it "rejects nonsense (3)" $ do
+      evaluate (parseExpression (lex "(4]")) `shouldThrow` anyException
+    -- it "rejects nonsense (4)" $ do
+    --   evaluate (parseExpression (lex "5 + * 5")) `shouldThrow` anyException
```
why exclude this?

lukasberglund replied:
This one is weird. When I run it, it throws an error, but somehow it still does not not pass this test.

ayazhafiz replied:
maybe the call to `evaluate` is causing this?

lukasberglund replied:
If I don't call evaluate it gives me this error during compilation: 
```
    • Couldn't match expected type ‘IO a0’
                  with actual type ‘Expression’
    • In the first argument of ‘shouldThrow’, namely
        ‘parseExpression (lex "5 + * 5")’
      In a stmt of a 'do' block:
        parseExpression (lex "5 + * 5") `shouldThrow` anyException
      In the second argument of ‘($)’, namely
        ‘do parseExpression (lex "5 + * 5") `shouldThrow` anyException’
   |
85 |       parseExpression (lex "5 + * 5") `shouldThrow` anyException
   |
```

lukasberglund commented:
> Hi Lukas, I think we will need to bump up the test coverage on this commit (it is rather low, at around 88%). Can you take a look at the coverage report on coveralls and add more tests to improve the coverage? Lmk if you need any help

I gotchu

lukasberglund commented:
@ayazhafiz What do you think about making a file for general helper functions like isAnyOf and splitWhen? What would you call such a file?

ayazhafiz commented:
Sounds great. call it Utils.

ayazhafiz commented:
Please rebase on master.

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,146 @@
+parseExpression' precidence tokens = (infixExpression, afterInfix)
+  where
+    (left, afterPrefix) = parsePrefix tokens
+    (infixExpression, afterInfix) = parseInfix precidence left afterPrefix
```
Please include this in the `where` clause of `parseExpression`.

lukasberglund replied:
I can't because I call this function from parseInfix.

ayazhafiz replied:
fair

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,146 @@
+    (infixExpression, afterInfix) = parseInfix precidence left afterPrefix
+
+parsePrefix :: [Token] -> (Expression, [Token])
+-- parsePrefix - Parse operations that do not depend on previous symbols.
```
```suggestion
-- Parse operations that do not depend on previous symbols.
```

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,146 @@
+    Grammar.Plus -> 1
+    Grammar.Div -> 2
+    Grammar.Times -> 2
+    Grammar.Mod -> 2
```
Consider moving this to the top of the parser module.

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,146 @@
+            prefix = Grammar.PrefixOperation operator expression
+
+parseInfix :: Int -> Expression -> [Token] -> (Expression, [Token])
+-- parseInfix - Given precidence and the left part of an operations parse all
```
```suggestion
-- Given precidence and the left part of an operations parse all
```

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,146 @@
+      [[]] -> []
+      _ -> map parseExpression splitTokens
+
+getClosingDelim :: Symbol.Symbol -> Symbol.Symbol
```
```suggestion
closingDelim :: Symbol.Symbol -> Symbol.Symbol
```

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,146 @@
+    Symbol.LParen -> Symbol.RParen
+    Symbol.StrBound -> Symbol.StrBound
+
+isSymbol :: Symbol.Symbol -> Token -> Bool
```
```suggestion
isOfSymbol :: Symbol.Symbol -> Token -> Bool
```

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,146 @@
+    Symbol.Mod -> Grammar.Mod
+    Symbol.RParen -> Grammar.Paren
+    Symbol.RBracket -> Grammar.Paren
+    _ -> error $ show sym
```
can you include a more descriptive error?

ayazhafiz commented on src/Parser.hs:
```diff
@@ -0,0 +1,146 @@
+  case sym of
+    Symbol.Plus -> Grammar.Positive
+    Symbol.Minus -> Grammar.Negative
+    _ -> error $ show sym
```
ditto

lukasberglund replied:
Why do you think it should go to the top of the file?

ayazhafiz commented on test/Parser/ParserSpec.hs:
```diff
@@ -0,0 +1,110 @@
+      parse (lex "1 + 2! foo()!") `shouldBe`
+        [InfixOperation (Number 1) Plus (Number 2), FunctionCall "foo" []]
+    -- it "rejects empty inputs" $ do
+    --   evaluate ( parse (lex "")) `shouldThrow` anyErrorCall
```
The problem is that this doesn't throw.

lukasberglund replied:
How do I make it throw?

ayazhafiz replied:
Change the parser so that it doesn't accept empty inputs. Or be okay with empty inputs.

lukasberglund replied:
This is another case of me throwing an error but it not really recognizing that I did throw an error.

ayazhafiz commented on test/Parser/ParserSpec.hs:
```diff
@@ -0,0 +1,110 @@
+    it "rejects nonsense (3)" $ do
+      evaluate (parseExpression (lex "(4]")) `shouldThrow` anyException
+    -- it "rejects nonsense (4)" $ do
+    --   parseExpression (lex "5 + * 5") `shouldThrow` anyException
```
This doesn't throw either. It parses the expression.

ayazhafiz commented on test/Parser/ParserSpec.hs:
```diff
@@ -0,0 +1,110 @@
+        PrefixOperation Negative (InfixOperation (Number 2) Minus (Number 2))
+      parseExpression (lex "[ +5]") `shouldBe`
+        PrefixOperation Positive (Number 5)
+    it "rejects nonsense (1)" $ do
```
Please rename all of these tests to better reflect what they are testing, and consider grouping them in "describe" blocks. For example, all parentheses tests could go under a `describe "parentheses"` block,

lukasberglund commented on src/Grammar.hs:
```diff
@@ -0,0 +1,46 @@
+  | String String
+  | Variable String
+  | FunctionCall String [Expression]
+  | Conditional
```
Would that be possible given that we have to define a return type for parseExpression?

ayazhafiz replied:
Yeah, you're right. Not sure why I thought that was possible

ayazhafiz commented on test/Parser/ParserSpec.hs:
```diff
@@ -0,0 +1,123 @@
+    it "rejects parentheses that don't match (2)" $ do
+      evaluate (parseExpression (lex "(4]")) `shouldThrow` anyException
+    -- it "rejects two infix operators following one another" $ do
+    --   parseExpression (lex "5 + * 5") `shouldThrow` anyException
```
can you add code so that the parser actually throws in this case?

lukasberglund replied:
I don't know how to make it actually throw
```haskell
error "some message"
```
doesn't  work apparently.

ayazhafiz commented:
That's weird. Can you show me the code you are using?

On Wed, Jan 22, 2020 at 8:22 PM lukasberglund <notifications@github.com>
wrote:

> *@lukasberglund* commented on this pull request.
> ------------------------------
>
> In test/Parser/ParserSpec.hs
> <https://github.com/ayazhafiz/hgb/pull/24#discussion_r369925853>:
>
> > +    -- it "rejects two infix operators following one another" $ do
> +    --   parseExpression (lex "5 + * 5") `shouldThrow` anyException
>
> I don't know how to make it actually throw
>
> error "some message"
>
> doesn't work apparently.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ayazhafiz/hgb/pull/24?email_source=notifications&email_token=AE6GL6Q4TRPICF6UVITXKKLQ7ELOXA5CNFSM4KJME3LKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCSXZKPA#discussion_r369925853>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE6GL6RTPZZZY6SI627EDZTQ7ELOXANCNFSM4KJME3LA>
> .
>

lukasberglund commented on src/Parser.hs:
```diff
@@ -0,0 +1,139 @@
+isOfSymbol :: Symbol -> Token -> Bool
+isOfSymbol sym token = (symbol token == sym)
+
+parse :: [Token] -> [Expression]
```
@ayazhafiz Would you want this to return something like [(Expression, Error)]?

ayazhafiz replied:
That sounds good to me. Maybe more like `([Expression], [Error])`.

lukasberglund commented on src/Parser.hs:
```diff
@@ -0,0 +1,139 @@
+  where
+    (result, []) = parseExpression' 0 xs
+
+parseExpression' :: Int -> [Token] -> (Expression, [Token])
```
@ayazhafiz If I have to add an error here, what would you suggest I make this function return?

lukasberglund replied:
On second thought I would probably just make it return (Expression, Error, [Token])

ayazhafiz commented:
How is it going with the errors? Can I help?

lukasberglund commented:
> How is it going with the errors? Can I help?

Yes! I started with something. It runs but it doesn't work for all types of errors yet and I haven't rewritten the tests to reflect the changes I made to the parser. 

It would be great if you told me if I am on the wrong track or whether I should change something about my approach. Getting hrer was surprisingly difficult!

ayazhafiz commented:
Okay, cool. I will take a look soon!

lukasberglund commented:
> For example, an InfixOperation should always be an `Expression Operator Expression`, not `(Maybe Expression) Operator (Maybe Expression)` (in the latter case, if the LHS was a Nothing, the whole expression wouldn't be an infix operation)! Something I would do is to try to record as much as possible in my AST before bailing out. For example, `1.2.3 + 5` could parsed something like `Number [1.] InfixOperator [Number [2.3] Plus Number [5]]` or `InfixOperator[Unknown [1.2.3] Plus Number [5]]`. This way, we preserve the structure of the grammar without having to have `Maybe`s, and can keep the errors independent of the AST.

Alright, I changed it up as per your recommendation. The only thing that is annoying now it that I created a seperate type of Blank datatype for each grammar structure (BlankOperator, BlankExpression, etc.) which doesn't seem very elegant.

> There are probably even better ways to approach this. Feel free to explore and implement how you see best or most interesting!
 :)

ayazhafiz commented:
> The only thing that is annoying now it that I created a seperate type of Blank datatype for each grammar structure (BlankOperator, BlankExpression, etc.) which doesn't seem very elegant.

Yeah, this is probably not the way we should go. I would recommend looking up how other parsers handle this (for example, check out Rust and TypeScript -- if you need help finding the sources, let me know and I'll point you to the relevant code). In cases where you need a blank operator or expression, we probably just shouldn't recover at all. It can be a best-effort kind of thing. But yeah something like `1 $$ 4` isn't parsable at all after `1 Invalid[$$] 4`.

lukasberglund commented:
> Yeah, this is probably not the way we should go. I would recommend looking up how other parsers handle this (for example, check out Rust and TypeScript -- if you need help finding the sources, let me know and I'll point you to the relevant code). In cases where you need a blank operator or expression, we probably just shouldn't recover at all. It can be a best-effort kind of thing. But yeah something like `1 $$ 4` isn't parsable at all after `1 Invalid[$$] 4`.

Could you point me to the relevant code in rust? I'm having a hard time grasping what's going on.

ayazhafiz commented:
Sure, I got you. Give me some time.

ayazhafiz commented:
@lukasberglund https://github.com/rust-lang/rustc-guide/blob/master/src/the-parser.md

lukasberglund commented:
I am having trouble understanding what Rust does, partially due to the fact I have never written in Rust. I am going to try to see what tutorials say on how to handle errors in parsers. Lmk if you think that is a bad idea or if there are tutorials you would recommend.

ayazhafiz commented:
Seems fine to me. Not really aware of any tutorials.

________________________________
От: lukasberglund <notifications@github.com>
Отправлено: Saturday, February 22, 2020 11:47:10 PM
Кому: ayazhafiz/hgb <hgb@noreply.github.com>
Копия: hafiz <ayaz.hafiz.1@gmail.com>; Mention <mention@noreply.github.com>
Тема: Re: [ayazhafiz/hgb] Added parser that handles math expressions, strings, variables, and function-calls. (#24)


I am having trouble understanding what Rust does, partially due to the fact I have never written in Rust. I am going to try to see what tutorials say on how to handle errors in parsers. Lmk if you think that is a bad idea or if there are tutorials you would recommend.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ayazhafiz/hgb/pull/24?email_source=notifications&email_token=AE6GL6W3IQ7U5R7NTNROQWLREISX5A5CNFSM4KJME3LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMVT6AA#issuecomment-590036736>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE6GL6VPRTH62ILNKDTKF4TREISX5ANCNFSM4KJME3LA>.

lukasberglund commented:
Okay so I changed things up so that now the parser gives up when it hits any parse errors. I think this makes sense since it makes the code a lot more readible and a lot of compilers seem to do it that way.

ayazhafiz commented on test/Lexer/LexerSpec.hs:
```diff
@@ -13,26 +13,32 @@ import Lexer (lex)
 import Symbol (Symbol(..))
 import Token (Token(..))
 
+lexRemoveEndOfFile :: String -> [Token]
```
just call this `driveLexer` or something and leave a comment that it removes EOF

ayazhafiz replied:
Bump for this

lukasberglund commented:
Hey mate. Can you take a look at coveralls and the tests? Ive tried a bunch of things but I can't get the coverage up

ayazhafiz commented:
Sure

________________________________
От: lukasberglund <notifications@github.com>
Отправлено: Wednesday, March 4, 2020 9:43:36 PM
Кому: ayazhafiz/hgb <hgb@noreply.github.com>
Копия: hafiz <ayaz.hafiz.1@gmail.com>; Mention <mention@noreply.github.com>
Тема: Re: [ayazhafiz/hgb] Added parser that handles math expressions, strings, variables, and function-calls. (#24)


Hey mate. Can you take a look at coveralls and the tests? Ive tried a bunch of things but I can't get the coverage up

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ayazhafiz/hgb/pull/24?email_source=notifications&email_token=AE6GL6W5GAZGMRZHQIB3GY3RF43YRA5CNFSM4KJME3LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEN3ZLHI#issuecomment-595039645>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE6GL6X23CC33LBPYY2TVGDRF43YRANCNFSM4KJME3LA>.

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,47 @@
+  = Operation
+      { name :: String
+      , arguments :: [Expression]
+      }
```
Is this supposed to represent a function-like operation (func [param1, param2]), a binary-like operation (lessThan [2, 3]), both, or something else? I think we can be more specific here.

ayazhafiz replied:
Hmm.. so do we permit pre and postfix function operations with no differentiation? E.g. you are suggesting 

```
func op1 op2
```

and

```
op1 + op2
```

are represented the same way internally (`Operation "func" [op1, op2]` and `Operation "+" [op1, op2]`).

This is a powerful pattern, but there are a couple of limitations:

- Complicates the parser (have multiple rules for the same output of "Operation" expression)
- Defers validation of `+` function to a later compiler stage. But at the parser level we have enough information to know that `+` can only ever have two operands, in an infix syntax.

What do you think?

lukasberglund replied:
> Hmm.. so do we permit pre and postfix function operations with no differentiation? E.g. you are suggesting func op1 op2 and op1 + op2 are represented the same way internally (Operation "func" [op1, op2] and Operation "+" [op1, op2]).

Yes I am suggesting they are represented the same way internally, although 
`func op1 op2`
would currently have to be written as 
`func(op1 op2)`

> Complicates the parser (have multiple rules for the same output of "Operation" expression)

Fair enough, if you want to, I can add InfixOperation and PrefixOperation again.

> Defers validation of + function to a later compiler stage. But at the parser level we have enough information to know that + can only ever have two operands, in an infix syntax.

Well the parser is currently written so that it will only parse infix operation that have to operands, so we are performing validation.

lukasberglund replied:
Lmk if I should add prefix and infix operations.

ayazhafiz replied:
How much do you think it will complicate our parser if we choose to use only one Operation?

Some other things to think about:

- How will we evaluate/transform these operations? For binary infix operations do we just assume a contract that there are only two parameters? This could lead to some assumptions in the state, whereas in a binary operation AST we can guarantee the number of parameters

- How will we express the name of an operation? User defined functions will have to be a string. Should built in infix operations be string defined too?

lukasberglund replied:
> How much do you think it will complicate our parser if we choose to use only one Operation?

I'm not sure I fully understand what you are asking. The parser is currently written to use only one operation. 

> How will we evaluate/transform these operations? For binary infix operations do we just assume a contract that there are only two parameters? This could lead to some assumptions in the state, whereas in a binary operation AST we can guarantee the number of parameters

Well it's true that we can't know for sure that there are two parameters in a sense (although the parser would never produce a binary operation that doesn't have more than two parameters). But isn't it the same for function calls? If the user defines a function that takes two parameters, we can't be sure that when he calls the function he will call it with two parameters.

> How will we express the name of an operation? User defined functions will have to be a string. Should built in infix operations be string defined too?

I think the best way to do this might be to edit the operation constructor to something like.
```haskell
Operation {opType :: OperationType, arguments :: [Expression]}
```
and define OperationType a bit like this
```haskell
data OperationType = Plus | Minus | ... | UserFunction String
```

ayazhafiz commented on test/Parser/ParserSpec.hs:
```diff
@@ -0,0 +1,171 @@
+      parseLineNoRemainder "(3 + 5) % 2!" `shouldBe`
+        Operation "Mod" [Operation "Plus" [Number 3, (Number 5)], Number 2]
+      parseLineNoRemainder "- [ 2-2 ]!" `shouldBe`
+        Operation "Minus" [Operation "Minus" [(Number 2), (Number 2)]]
```
Looks like we are recording Operators as operations rather via the Operator grammar. Why is this? Should Operation be updated to take in Operator as the name? In that case, we should also distinguish between e.g. BinaryOperations as in most of these cases...

ayazhafiz replied:
Why not

```haskell
Operation
   { name :: Operator
    , arguments :: [Expression]
    }
```

ayazhafiz replied:
Well, unless we decide that this grammar (`Operation`) should only be for functions. Then we should probably keep as a string.

ayazhafiz commented:
Actually, weird that coveralls didn't pick up the new coverage changes too much... on my machine it's up to 98% now. I will investigate later, have to go to work rn...

ayazhafiz commented on src/Error.hs:
```diff
@@ -5,7 +5,7 @@ import Token (Token(..))
 
 data Error =
   Error ErrorType Token
-  deriving (Show, Read, Eq)
```
Why derive it if we don't need it? That is part of what coverage helps us catch.

lukasberglund replied:
True, I just figured we are making our datatype less useful than it could be. But I guess we can always add it later.

ayazhafiz commented:
Okay, sounds good to me. Let’s try it out! Can you fix with changes regarding OperationType?

Another option regarding OperationType is just keep the name as a string. Then we can an expression like 1+2 parsed as an operation with name “+”. IMO this might be better so that we can perform lookup in a function table in a consist way later on.

________________________________
От: lukasberglund <notifications@github.com>
Отправлено: Friday, March 6, 2020 6:50:31 PM
Кому: ayazhafiz/hgb <hgb@noreply.github.com>
Копия: hafiz <ayaz.hafiz.1@gmail.com>; Mention <mention@noreply.github.com>
Тема: Re: [ayazhafiz/hgb] Added parser that handles math expressions, strings, variables, and function-calls. (#24)


@lukasberglund commented on this pull request.

________________________________

In src/Grammar.hs<https://github.com/ayazhafiz/hgb/pull/24#discussion_r389221035>:

> +  = Operation
+      { name :: String
+      , arguments :: [Expression]
+      }


How much do you think it will complicate our parser if we choose to use only one Operation?

I'm not sure I fully understand what you are asking. The parser is currently written to use only one operation.

How will we evaluate/transform these operations? For binary infix operations do we just assume a contract that there are only two parameters? This could lead to some assumptions in the state, whereas in a binary operation AST we can guarantee the number of parameters

Well it's true that we can't know for sure that there are two parameters in a sense (although the parser would never produce a binary operation that doesn't have more than two parameters). But isn't it the same for function calls? If the user defines a function that takes two parameters, we can't be sure that when he calls the function he will call it with two parameters.

How will we express the name of an operation? User defined functions will have to be a string. Should built in infix operations be string defined too?

I think the best way to do this might be to edit the operation constructor to something like.

Operation {opType :: OperationType, arguments :: [Expression]}

and define OperationType a bit like this

data OperationType = Plus | Minus | ... | UserFunction String

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ayazhafiz/hgb/pull/24?email_source=notifications&email_token=AE6GL6UK47J5URVX7OO5IYDRGGY7PA5CNFSM4KJME3LKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCYMLA2I#discussion_r389221035>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE6GL6XULNHTGKW3MMZKHCLRGGY7PANCNFSM4KJME3LA>.

ayazhafiz commented:
Okay, sounds good to me. Let’s try it out! Can you fix with changes regarding OperationType?

Another option regarding OperationType is just keep the name as a string. Then we can an expression like 1+2 parsed as an operation with name “+”. IMO this might be better so that we can perform lookup in a function table in a consist way later on.

________________________________
От: lukasberglund <notifications@github.com>
Отправлено: Friday, March 6, 2020 6:50:31 PM
Кому: ayazhafiz/hgb <hgb@noreply.github.com>
Копия: hafiz <ayaz.hafiz.1@gmail.com>; Mention <mention@noreply.github.com>
Тема: Re: [ayazhafiz/hgb] Added parser that handles math expressions, strings, variables, and function-calls. (#24)


@lukasberglund commented on this pull request.

________________________________

In src/Grammar.hs<https://github.com/ayazhafiz/hgb/pull/24#discussion_r389221035>:

> +  = Operation+      { name :: String+      , arguments :: [Expression]+      }

How much do you think it will complicate our parser if we choose to use only one Operation?

I'm not sure I fully understand what you are asking. The parser is currently written to use only one operation.

How will we evaluate/transform these operations? For binary infix operations do we just assume a contract that there are only two parameters? This could lead to some assumptions in the state, whereas in a binary operation AST we can guarantee the number of parameters

Well it's true that we can't know for sure that there are two parameters in a sense (although the parser would never produce a binary operation that doesn't have more than two parameters). But isn't it the same for function calls? If the user defines a function that takes two parameters, we can't be sure that when he calls the function he will call it with two parameters.

How will we express the name of an operation? User defined functions will have to be a string. Should built in infix operations be string defined too?

I think the best way to do this might be to edit the operation constructor to something like.

Operation {opType :: OperationType, arguments :: [Expression]}

and define OperationType a bit like this

data OperationType = Plus | Minus | ... | UserFunction String

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ayazhafiz/hgb/pull/24?email_source=notifications&email_token=AE6GL6UK47J5URVX7OO5IYDRGGY7PA5CNFSM4KJME3LKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCYMLA2I#discussion_r389221035>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE6GL6XULNHTGKW3MMZKHCLRGGY7PANCNFSM4KJME3LA>.

lukasberglund commented:
> Okay, sounds good to me. Let’s try it out! Can you fix with changes regarding OperationType? Another option regarding OperationType is just keep the name as a string. Then we can an expression like 1+2 parsed as an operation with name “+”. IMO this might be better so that we can perform lookup in a function table in a consist way later on.

I did what you suggested as an alternative option!

ayazhafiz commented on src/Grammar.hs:
```diff
@@ -0,0 +1,47 @@
+  | Negative
+  | Positive
+  | Paren
+  deriving (Show, Eq)
```
Will you get rid of this since we are moving to all functions having string names?

lukasberglund commented:
It looks like all that's left is to bump up the test coverage. Could you take a look at that?

ayazhafiz commented:
Sure

On Sat, Mar 7, 2020 at 5:05 PM lukasberglund <notifications@github.com>
wrote:

> It looks like all that's left is to bump up the test coverage. Could you
> take a look at that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ayazhafiz/hgb/pull/24?email_source=notifications&email_token=AE6GL6WEQ37KLZW4JEBFDRTRGLVNXA5CNFSM4KJME3LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOEJOEY#issuecomment-596154131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE6GL6VSRWQFELPA55W5NUDRGLVNXANCNFSM4KJME3LA>
> .
>

ayazhafiz commented:
Hey Lukas, looks like the git history for this branch is pretty messed up (with branching). This makes rebasing pretty difficult. In the future let's take care to keep history on one branch so we don't have to merge branches and can rebase instead.

ayazhafiz commented:
@lukasberglund looks like the error to catch missing expression endings is not working correctly:

```
    it "should expect ! to end expressions" $ do
      parseError "1+1" `shouldBeError` ExpectedSymbol Symbol.ExprEnd

--- 

  1) Parser.Parser.errors should expect ! to end expressions
       expected: ExpectedSymbol ExprEnd
        but got: ExpectedOperator
```

Do you want to fix this? Or should I give it a shot

lukasberglund commented:
> Hey Lukas, looks like the git history for this branch is pretty messed up (with branching). This makes rebasing pretty difficult. In the future let's take care to keep history on one branch so we don't have to merge branches and can rebase instead.

Sorry about that. I'm not quite sure what it means to keep history on one branch. What should I do/ not do to make sure that happens?

Also, I'll fix the error.

lukasberglund commented:
Okay I've simplified the errors let me know what you think. I changed the code of the parser to take advantage of the fact that Either is a monad (I just learned about Monads!) I wanted to put that in a later commit but somehow it snuck it's way into this one. I'm gonna make it a bit more readible in the next half hour so maybe read the code then.

ayazhafiz commented:
Yay!

________________________________
От: lukasberglund <notifications@github.com>
Отправлено: Saturday, March 7, 2020 9:53:13 PM
Кому: ayazhafiz/hgb <hgb@noreply.github.com>
Копия: hafiz <ayaz.hafiz.1@gmail.com>; Mention <mention@noreply.github.com>
Тема: Re: [ayazhafiz/hgb] Added parser that handles math expressions, strings, variables, and function-calls. (#24)


Okay I've simplified the errors let me know what you think. I changed the code of the parser to take advantage of the fact that Either is a monad (I just learned about Monads!) I wanted to put that in a later commit but somehow it snuck it's way into this one. I'm gonna make it a bit more readible in the next half hour so maybe read the code then.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ayazhafiz/hgb/pull/24?email_source=notifications&email_token=AE6GL6TP6CSCMJAARWNV4J3RGMXETA5CNFSM4KJME3LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOENJWA#issuecomment-596169944>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE6GL6UGA7JIHWOQ5ENVS7LRGMXETANCNFSM4KJME3LA>.

lukasberglund commented:
@ayazhafiz  so right now the ordering of our functions is a bit weird. First we have simple mapping functions then we go from high-level to low level. Would it be more true to convention and better to just go from most low level at the top to most high level at the bottom?

ayazhafiz commented:
> @ayazhafiz so right now the ordering of our functions is a bit weird. First we have simple mapping functions then we go from high-level to low level. Would it be more true to convention and better to just go from most low level at the top to most high level at the bottom?

Yes

lukasberglund commented on src/Parser.hs:
```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE MultiWayIf #-}
```
Not sure how it works but it allows for the use of guards (these guys --> '|') in a do structure through this syntax:

```haskell
do
...
if
 | a == b -> 1
 | otherwise -> 2
```
or something like that

ayazhafiz replied:
Interesting. I like it. Much pattern match.

lukasberglund commented on src/Parser.hs:
```diff
@@ -0,0 +1,176 @@
+--    operator that was passed in as an argument. That means, if it encounters
+--    an operator that is lower precidence than what it is parsing it stops.
+--    As I am writing this even I am not quite sure why this works but somehow
+--    through the power of recursion everything turns out right.
```
That's a pretty good explanation, thanks!

lukasberglund commented:
> > @ayazhafiz so right now the ordering of our functions is a bit weird. First we have simple mapping functions then we go from high-level to low level. Would it be more true to convention and better to just go from most low level at the top to most high level at the bottom?
> 
> Yes

I'll do that then.

lukasberglund commented on src/Parser.hs:
```diff
@@ -0,0 +1,176 @@
+
+parseExpression' :: Int -> [Token] -> Either Error (Expression, [Token])
+parseExpression' precidence tokens = do
+  (left, afterLeft) <- parsePrefix tokens
```
A do statement allows simplified syntax for monadic structures. 

`Either a b` is a monad with the value that's wrapped in the monad being whatever is on the right (In this case that's `(Expression, [Token])`).  The do statement allows me to  use statements like 
```haskell
(left, afterLeft) <- parsePrefix tokens
```
to extract the values out of monads without having to deal with failure (in this case an `Error` value).  In this example if `parsePrefix tokens` returns an `Right (Expression, [Token])` value as expected it gets assigned to `(left, afterLeft)`. If `parsePrefix tokens` returns an error then parseExpression' stops and just returns that error.

An alternative way of writing  this would be 
```haskell
parseExpression' precidence tokens = 
    parsePrefix tokens >>=
        (left, afterLeft) -> ...
```
Or if you haven't learned about monads yet, you could just write
```haskell
parseExpression' precidence tokens = 
    case parsePrefix tokens of
        Left e -> Left e
        Right (left, afterLeft) -> ...
```

ayazhafiz replied:
In Haskell, do all monads have a `Left` and `Right`? Or is this that just how `Error` implements a monad?

Is it correct that `<-` is just syntatic sugar for `>>=` with the additional benefit that it bring `left` and `afterLeft` into scope as variables?

lukasberglund replied:
It might help to skim [learnyouahaskell.com/a-fistful-of-monads](url) to see how monads work in haskell. 

> In Haskell, do all monads have a Left and Right? Or is this that just how Error implements a monad?

Not quite. In Haskell, `Either a b` is a monad, but `IO` and `Maybe`, and even lists are monads too.  For example if `parseExpression'` and `parsePrefix` returned the monad `Maybe (Expression, [Token])`, then 
```haskell
(left, afterLeft) <- parsePrefix tokens
```
would do the same thing if `parsePrefix tokens` returned `Just (Expression, [Token])` but if `parsePrefix tokens` returned `Nothing` then all of `parseExpression`  returns `Nothing`. 

> Is it correct that <- is just syntatic sugar for >>= with the additional benefit that it bring left and afterLeft into scope as variables?

Yes, it's just syntactic sugar for `>>=` but there is not additional benefit. `>>=` is an infix operator that takes a monad and a function that returns a monad and returns a monad so it's declaration is something like.
```haskell
(>>=) :: m a -> (a -> m b) -> m b
```
 Thus by writing

```haskell
parseExpression' precidence tokens = 
    parsePrefix tokens >>=
        \(left, afterLeft) -> ...
```
You can finish writing what's behind the error as the rest of your function definition with `(left, afterLeft)` in scope. However this is kind of awkward because you have to define a lambda inside of your function.

lukasberglund commented on src/Parser.hs:
```diff
@@ -0,0 +1,176 @@
+parseInfix precidence left tokens@(operatorTok:afterOperator)
+  | terminatesExpressions operatorSym = Right (left, tokens)
+  | not (isInfixOperator operatorSym) =
+    Left $ Error (Expected "Operator or '!'") operatorTok
```
Hindent actually did this.

ayazhafiz replied:
Sounds good
